From 6807bf4e0befe12552a117ad46005d0969ee4d73 Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Thu, 14 Mar 2024 05:24:29 +0200
Subject: [PATCH] [FSW-402] fs1: bootrom_init: add iccm init

This patch adds ARC1 ICCM init functionality for IO die 0
with TODO added for memcpy block.

Change-Id: If81db66fa2b86ed72a02bc8f4c8bb8e0b06c32f6
---

diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index 11bd902..6ff7509 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -105,3 +105,33 @@
 	boot_rom_global.validation_pass = cold_rst_data.validation_flag;
 	boot_rom_global.efuse_read = cold_rst_data.efuse_read_flag;
 }
+
+/* TODO: Add functionality once mem init feature is added to memcpy block */
+int boot_rom_mem_init(__unused uint64_t addr, __unused uint32_t size)
+{
+	/* Empty Body */
+	return 0;
+}
+
+void boot_rom_iccm_init(void)
+{
+	struct fs1_cold_rst_data cold_rst_data;
+	int ret;
+
+	if (boot_rom_global.die_type != IO_DIE_MASTER)
+		return;
+
+	/* Perform arc1 iccm init only on cold reset */
+	cold_rst_data.data = regmap_readl(IO0_HALF0_PSOC_BOOT(mmPSOC_BOOT_CONF_COLD_RST_REG_2));
+	if (cold_rst_data.arc1_iccm_init_done)
+		return;
+
+	ret = boot_rom_mem_init(HL_ARC1_ICCM_FULL_BASE, ARC1_ICCM_SIZE);
+	if (ret) {
+		boot_rom_status_update(PROGRESS_INIT_ARC1_ICCM_INIT_FAILED);
+		return;
+	}
+
+	cold_rst_data.arc1_iccm_init_done = 1;
+	regmap_writel(cold_rst_data.data, IO0_HALF0_PSOC_BOOT(mmPSOC_BOOT_CONF_COLD_RST_REG_2));
+}
diff --git a/common/common.c b/common/common.c
index 9bb2ba4..16484ac 100644
--- a/common/common.c
+++ b/common/common.c
@@ -32,5 +32,7 @@
 
 	boot_rom_read_cold_flops();
 
+	boot_rom_iccm_init();
+
 	spi_flash_init();
 }
diff --git a/common/common.h b/common/common.h
index dfb5ed6..b0cf8ca 100644
--- a/common/common.h
+++ b/common/common.h
@@ -12,5 +12,6 @@
 void boot_rom_clear_status(void);
 void boot_rom_persistent_dccm_init(uint8_t die_id);
 void boot_rom_read_cold_flops(void);
+void boot_rom_iccm_init(void);
 
 #endif /* _COMMON_H_ */
